### ISSUE TITLE
Type mismatch between Getter/Fold and to in 0.4.2. (Star Const -> Forget related) 

### ISSUE THREAD

**Issue by benkolera** [purescript-contrib/purescript-profunctor-lenses#24]
```
newtype XPathPoint = XPathPoint
  { selector :: String
  , offset   :: Int
  }

selector :: GetterP XPathPoint String
selector = to $ \ (XPathPoint s) -> s.selector
```

Gives this compile error:

```
at /Users/bkolera/src/benaud/src/CommentManager/Types.purs line 48, column 12 - line 50, column 1

  Could not match type

    Forget

  with type

    Star


while trying to match type Forget
  with type Star
while checking that expression (($) to) (\_17 ->
                                           case _17 of
                                             (XPathPoint s) -> ...
                                        )
  has type Forget String String String -> Forget String XPathPoint XPathPoint
in value declaration selector
```

I note that Fold changed from 

```
type Fold r s t a b = Optic (Star (Const r)) s t a b
```

in 0.3.5 to this in 0.4.x (I'm using 0.4.2)

```
type Fold r s t a b = Optic (Forget r) s t a b
```

Which is the source of the type error, but don't quite have a clue whether the library is actually broken in 0.4.0 or I need to do something slightly differently.

**Comment by benkolera** [comment]
Appears to have broken by c58fde5671df9c5e06cb70e90ba302acf16cbf56

**Comment by benkolera** [comment]
Take a look at https://github.com/benkolera/purescript-profunctor-lenses/commit/adc1a768148911cd568d725b0bb91648d5b87fc3 . I don't suggest that this is the answer as it is pretty ugly, but it does work...

**Comment by clayrat** [comment]
Pinging @zrho, as he's the author of the commit in question.

**Comment by zrho** [comment]
I see, I forgot to change some functions on the switch. I've tinkered a bit, and came up with this: https://github.com/zrho/purescript-profunctor-lenses/commit/f3a83d3d92e8adeab9cb4789b4407ba5a60fcf50

It passed the limited amount of tests I threw at it; would it solve your problem?

Probably, it is worthwhile to restrict the type of `to` to only include `Fold`s, since using `lmap` without an `rmap` in its company messes things up if the result is not forgot, i.e. the optic is not used for folding.

**Comment by benkolera** [comment]
That looks like it should solve the issue and make sense to me. I need to level up my thinking with profunctors, it seems. :)

**Comment by benkolera** [comment]
@zrho Ahh, no, that doesn't quite work, sorry. It's because the optic that powers to is 

```
Optic p s t a t 
```

Which is incompatible with GetterP, which as a user of the library this feels wrong to me. With the example from last time:

```
  Could not match type
    String
  with type
    XPathPoint

while trying to match type _0 _1 _2
  with type Forget String XPathPoint XPathPoint
while checking that expression (($) to) (\$17 ->
                                           case $17 of
                                             (XPathPoint s) -> ...
                                        )
  has type Forget String String String -> Forget String XPathPoint XPathPoint
in value declaration selector

where _0 is an unknown type
      _2 is an unknown type
      _1 is an unknown type
```

I may just be a noob, but I don't see how you'd be able to implement to properly with just profunctors alone. Because the value that we need to set is actually the r of the Forget data type which is hidden from the profunctor by design. If we aren't setting the value that we want into that r of the forget, then view isn't going to work how we want it to.

If you look at this commit https://github.com/benkolera/purescript-profunctor-lenses/commit/e628886f6d84877c1da4f7eb8ca42c43fe637c63 you'll see where your solution isn't compiling and how just constructing a simple Forget value with the output remembered falls over when we use the getter with preview, because unlike with Star there is no inner functor for First to hide inside forget. 

Again I'm a noob at this stuff, but the only way that I see this working with view and preview as we'd expect is to  make to an optic over the Forget profunctor, but with a Contravariant functor inside (with view using Identity) so that we can create an optic that can go through either function and return what we want. 

You may already know this or have a better way that I'm not thinking about, though. I think that I'm mainly trying to speak out loud here in hopes that I may learn something in the process. :)

**Comment by zrho** [comment]
tldr: `Getter` isn't meant to be used that way; in the future, documentation or better naming should make that clear. 

Ah, I see that this could be a bit misleading, the way things are named, since the documentation is partly missing and partly just wrong, because the concepts are still evolving a bit.

`Getter`s are (at least in my head) defined to be the things functions like `view` consume. The value that is contained and accumulated by `Forget` in a chain of optics is the end-result of such a chain that is to be gotten. They aren't really meant to be used in this way on the definition side.

This is why your example fails at the usage site: preview wants to get a `First Int` as the result of the fold, but by providing a `GetterP GetterTest Int`, you tell him that he can only have `Int`s, since a `GetterP GetterTest Int` is a thing specifically meant for `view` to consume, which wants to have an `Int`.

At the declaration side, it fails because of the following reason: the type `Fold r s t a b` basically corresponds to a function `(a -> r) -> s -> r`, a `foldMap` basically. Note that the types `t` and `b` are irrelevant. This is why, operationally, merely using `lmap` works. By declaring a `Fold Int GetterTest GetterTest Int Int` (i.e. a `GetterP GetterTest Int`) you force the second `GetterTest` to change into `Int`, which it doesn't with `lmap`, but could, because the `t` and `b` are ignored anyways. Here, I agree, that it would be worthwhile to restrict this to a `Fold`, but for a different reason: this would emphasize the way the types should be used.

The result would look like this:

```
to :: forall r s t a b. (s -> a) -> Fold r s t a b
to f p = Forget (runForget p <<< f)
```

Note that the type `r` is independent of `s` and `a`, because `to` could be used in the middle of a chain of optics, but `r` is the end-result of a fold. Your version of `to` fails in that regard, because it fixes `r` to be `a`, and because it ignores the rest of the chain.

**Comment by benkolera** [comment]
Ahh yeah, got it. I knew there was something very important missing from mine. That clears up where I was getting so confused, so thank you! :)

Hmm, coming from Control.Lens in haskell, not being able to use a getter with preview is _very_ jarring.

I often find myself doing something like this in haskell:

foo ^? aLens . _Just . to show
foo ^? bLens . to show

As being able to just to tack on some kind of function onto the end of the chain is quite handy.

**Comment by zrho** [comment]
Well, you can use a getter, but you can't use a `Getter`. Just try those
two exact examples; they do work, because `to` specializes to a `Fold`.

**Comment by benkolera** [comment]
Oh, yeah. Got it. Pretty confusing if you don't know the underlaying types rather than just treating things as Lenses/Prisms/Traversals/Isos/Getters with the aliases, but not a deal breaker and should be fine with some extra doco.

Sorry about me being dense there. It's very early in the morning here. :)

Thanks for your help!

**Comment by paf31** [comment]
@zrho Could you please make that change into a PR?

**Comment by paf31** [comment]
@zrho With or without the change, this fails to type-check, which I find odd:

``` purescript
getter :: forall t r a s. (Traversable t, Monoid r) => (s -> t a) -> FoldP r s a
getter f = to f <<< traversed
```

**Comment by paf31** [comment]
A simpler example:

``` purescript
string :: forall r. FoldP r F.Foreign String
string = to F.readString <<< traversed
```

gives

``` text
  Could not match type

    Either ForeignError String

  with type

    Foreign


while trying to match type _0 Foreign _1
  with type Forget r0 Foreign Foreign
while checking that expression ((<<<) (to readString)) traversed
  has type Forget r0 String String -> Forget r0 Foreign Foreign
in value declaration string
```

**Comment by zrho** [comment]
That is strange. After the change, i.e. with the version from the PR,

``` purescript
main = do
  print $ F.toForeign "Foo" ^.. string
  print $ Tuple [1, 2, 3] "Bar" ^.. getter (\(Tuple a _) -> a)

getter :: forall t r a s. (Traversable t, Monoid r) => (s -> t a) -> FoldP r s a
getter f = to f <<< traversed

string :: forall r. (Monoid r) => FoldP r F.Foreign String
string = to F.readString <<< traversed
```

compiles just fine and gives the desired output:

```
Cons ("Foo") (Nil)
Cons (1) (Cons (2) (Cons (3) (Nil)))
```

**Comment by paf31** [comment]
I just pulled the PR branch and I still have the same issue. I must be doing something wrong, I'll look again later.

**Comment by MonoidMusician** [comment]
The PR was merged (#27) so I think this can be closed now?

**Comment by paf31** [comment]
Yes, thanks,
